core/src/jwk.rs

```diff
@@ -0,0 +1,180 @@
+        jwk.key_id = Some(did.id.clone());
+        Ok(Self(jwk))
+    }
+
+    pub fn with_private_key(self, pk: &str) -> anyhow::Result<Self> {
+        let pk = hex::decode(pk)?;
+        let params = match self.0.params.clone() {
+            Params::OKP(mut op) => {
+                op.private_key = Some(Base64urlUInt(pk));
+                Params::OKP(op)
+            }
+            Params::EC(mut ec) => {
+                ec.ecc_private_key = Some(Base64urlUInt(pk));
+                Params::EC(ec)
+            }
+            _ => anyhow::bail!("Unsupported JWK Params"),
+        };
+        let mut jwk = JWK::from(params);
+        jwk.key_id = self.key_id.clone();
+        Ok(Self(jwk))
+    }
+}
+
+impl From<JWK> for Jwk {
+    fn from(value: JWK) -> Self {
+        Self(value)
+    }
+}
+
+impl Deref for Jwk {
+    type Target = JWK;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+async fn did_as_jwk(resolver: &dyn DIDResolver, id: &str) -> anyhow::Result<Option<JWK>> {
+    let (res, object, _) = dereference(resolver, id, &DereferencingInputMetadata::default()).await;
+    if res.error.is_none() {
+        match object {
+            Content::Object(Resource::VerificationMethod(vm)) => {
+                let jwk = vm.get_jwk()?;
+                return Ok(Some(jwk));
+            }
+            Content::DIDDocument(doc) => {
+                if let Some(vms) = &doc.verification_method {
+                    for vm in vms {
+                        if let VerificationMethod::Map(vm) = vm {
+                            if let Ok(jwk) = vm.get_jwk() {
+                                return Ok(Some(jwk));
+                            }
+                        }
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+    Ok(None)
+}
+
+fn pk_as_jwk(public_key: &str) -> anyhow::Result<JWK> {
+    let (_, data) = multibase::decode(public_key)?;
+    let curve = match data[0..2] {
+        [0xed, 0x01] => "Ed25519",
+        [0xe7, 0x01] => "SECP256K1",
+        _ => anyhow::bail!("Unknown encoding prefix"),
+    };
+    let mut jwk = JWK::from(Params::OKP(OctetParams {
+        curve: curve.to_string(),
+        public_key: Base64urlUInt(data[2..].to_vec()),
+        private_key: None,
+    }));
+    jwk.key_id = Some(public_key.to_string());
+    Ok(jwk)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ssi::did::DIDMethod;
+    use ssi::did::VerificationMethod;
+
+    #[tokio::test]
+    async fn should_convert_did_key_generated_without_vm() {
+        let jwk = JWK::generate_ed25519().unwrap();
+        let did = did_method_key::DIDKey
+            .generate(&ssi::did::Source::Key(&jwk))
+            .unwrap();
+        let did = DidDocument::new(&did);
+        let other_jwk = Jwk::new(&did).await.unwrap();
+        if let (Params::OKP(did), Params::OKP(orig)) = (&other_jwk.params, jwk.params) {
+            assert_eq!(did.public_key, orig.public_key)
+        }
```

Comment:
Should we have an else `panic` branch for the failure case when the let expressions do not match?